rows_test.go

```diff
@@ -13,3 +13,99 @@
 // limitations under the License.
 
 package spannerdriver
+
+import (
+	"database/sql/driver"
+	"fmt"
+	"io"
+	"testing"
+
+	"cloud.google.com/go/spanner"
+	sppb "google.golang.org/genproto/googleapis/spanner/v1"
+)
+
+type testIterator struct {
+	metadata *sppb.ResultSetMetadata
+	rows     []*spanner.Row
+	index    int
+}
+
+func (t *testIterator) Next() (*spanner.Row, error) {
+	if t.index == len(t.rows) {
+		return nil, io.EOF
+	}
+	row := t.rows[t.index]
+	t.index++
+	return row, nil
+}
+
+func (t *testIterator) Stop() {
+}
+
+func (t *testIterator) Metadata() *sppb.ResultSetMetadata {
+	return t.metadata
+}
+
+func newRow(t *testing.T, cols []string, vals []interface{}) *spanner.Row {
+	row, err := spanner.NewRow(cols, vals)
+	if err != nil {
+		t.Fatalf("failed to create test row: %v", err)
+	}
+	return row
+}
+
+func TestRows_Next(t *testing.T) {
+	cols := []string{"COL1", "COL2", "COL3"}
+	it := testIterator{
+		metadata: &sppb.ResultSetMetadata{
+			RowType: &sppb.StructType{
+				Fields: []*sppb.StructType_Field{
+					{Name: "COL1", Type: &sppb.Type{Code: sppb.TypeCode_INT64}},
+					{Name: "COL2", Type: &sppb.Type{Code: sppb.TypeCode_STRING}},
+					{Name: "COL3", Type: &sppb.Type{Code: sppb.TypeCode_FLOAT64}},
+				},
+			},
+		},
+		rows: []*spanner.Row{
+			newRow(t, cols, []interface{}{int64(1), "Test1", 3.14 * 1.0}),
+			newRow(t, cols, []interface{}{int64(2), spanner.NullString{}, spanner.NullFloat64{}}),
+			newRow(t, cols, []interface{}{int64(3), "Test3", 3.14 * 3.0}),
+		},
+	}
+
+	rows := rows{it: &it}
+	dest := make([]driver.Value, 3)
+	values := make([][]driver.Value, 0)
+	for {
+		err := rows.Next(dest)
+		if err == io.EOF {
+			break
+		}
+		row := make([]driver.Value, 3)
+		copy(row, dest)
+		values = append(values, row)
+	}
+	if g, w := len(values), 3; g != w {
+		t.Fatalf("values length mismatch\nGot: %v\nWant: %v", g, w)
+	}
+	for i := int64(1); i <= int64(len(values)); i++ {
```

Comment:
nits: can we start with 0? I was a bit confused when I first looked at this.

Reply:
Done